File: c++/0003-longest-substring-without-repeating-characters.cc
#include <gtest/gtest.h>

#include <string_view>
#include <unordered_map>

namespace longestSubstringWithoutRepeatingCharacters {

int solve(const std::string_view& s) {
  if (s.empty()) {
    return 0;
  }

  int max_len = 0;

  // start if the offset at which we last saw a repeated character
  int start = 0;
  std::unordered_map<char, int> char_positions;

  for (int i = 0; i < s.length(); ++i) {
    auto c = s[i];
    auto it = char_positions.find(c);
    // have we seen c since the last repeated character?
    if (it != char_positions.end() && it->second >= start) {
      start = it->second + 1;
    }

    char_positions[c] = i;

    max_len = std::max(max_len, i - start + 1);
  }

  return max_len;
}

}  // namespace longestSubstringWithoutRepeatingCharacters

TEST(LongestSubstringWithoutRepeatingCharacters, TestCases) {
  EXPECT_EQ(longestSubstringWithoutRepeatingCharacters::solve("abcabcbb"), 3);
  EXPECT_EQ(longestSubstringWithoutRepeatingCharacters::solve("bbbbb"), 1);
  EXPECT_EQ(longestSubstringWithoutRepeatingCharacters::solve("pwwkew"), 3);
}

File: swift/Sources/LeetCode/0003LongestSubstringWithoutRepeatingCharacters.swift

public func longestSubstringWithoutRepeatingCharacters(_ s: String) -> Int {
    guard !s.isEmpty else {
        return 0
    }
    
    var maxLength = 0
    var start = 0
    var charPositions = Dictionary<Character, Int>()
    
    for (index, char) in s.enumerated() {
        let pos = charPositions[char, default: -1]
        if (pos != -1 && pos >= start) {
            start = pos + 1
        }
        
        charPositions[char] = index
        
        maxLength = max(maxLength, index - start + 1)
    }
    
    return maxLength
}
